client/src/components/ThesesGanttChart/ThesesGanttChart.tsx

```diff
@@ -1,5 +1,153 @@
+import { useMemo, useState } from 'react'
+import { formatDate, formatUser } from '../../utils/format'
+import { ThesisStateColor } from '../../config/colors'
+import ThesisPreviewModal from '../ThesisPreviewModal/ThesisPreviewModal'
+import { IThesis, ThesisState } from '../../requests/responses/thesis'
+import { Center, Group, Stack, Text } from '@mantine/core'
+import LabeledItem from '../LabeledItem/LabeledItem'
+import ThesisStateBadge from '../ThesisStateBadge/ThesisStateBadge'
+import { Presentation } from 'phosphor-react'
+import { arrayUnique } from '../../utils/array'
+
 const ThesesGanttChart = () => {
-  return <></>
+  const { theses } = useThesesContext()
+
+  const [openedThesis, setOpenedThesis] = useState<IThesis>()
+
+  const data = useMemo<IGanttChartDataElement[] | undefined>(() => {
+    if (!theses) {
+      return undefined
+    }
+
+    const result: IGanttChartDataElement[] = []
+
+    for (const thesis of theses.content) {
+      const getAdjustedEndDate = (state: ThesisState, startDate: Date, endDate: Date) => {
+        if (state === ThesisState.PROPOSAL && state === thesis.state) {
+          // Proposal phase should be at least 4 weeks long if not completed yet
+          return new Date(Math.max(endDate.getTime(), startDate.getTime() + 3600 * 24 * 28 * 1000))
+        }
+
+        if (state === ThesisState.WRITING && state === thesis.state && thesis.endDate) {
+          // Writing phase should end at endDate if not completed yet
+          return new Date(thesis.endDate)
+        }
+
+        return endDate
+      }
+
+      result.push(
+        ...thesis.advisors.map((advisor) => ({
+          id: thesis.thesisId,
+          groupId: advisor.userId,
+          groupName: formatUser(advisor),
+          columns: [
+            thesis.students.map((user) => formatUser(user, { withUniversityId: false })).join(', '),
+            thesis.title,
+          ],
+          timeline: thesis.states.map((state) => ({
+            startDate: new Date(state.startedAt),
+            endDate: getAdjustedEndDate(
+              state.state,
+              new Date(state.startedAt),
+              new Date(state.endedAt),
+            ),
+            color: ThesisStateColor[state.state],
+          })),
+          events: thesis.presentations.map((presentation) => ({
+            icon: <Presentation />,
+            time: new Date(presentation.scheduledAt),
+          })),
+        })),
+      )
+    }
+
+    return result
+  }, [theses])
+
+  const visibleStates: ThesisState[] = theses
+    ? arrayUnique<ThesisState>([
+        ...theses.content.reduce<ThesisState[]>(
+          (prev, curr) => [
+            ...prev,
+            ...curr.states.filter((row) => row.startedAt !== row.endedAt).map((row) => row.state),
+          ],
+          [],
+        ),
+      ])
+    : []
+
+  return (
+    <div>
+      <GanttChart
+        columns={[
+          { label: 'Student', width: '120px', textAlign: 'center' },
```

Comment:
Would it make sense to switch to relative values like viewport width instead of sticking to the static pixel values?

Reply:
Mmh this makes it quite small on mobile. I will change it to a percentage